csrc/scheduler/matmul.cpp

```diff
@@ -49,6 +55,386 @@ void moveInnerBroadcastLeft(TensorView* tv, int number_of_inner_pos = 3) {
+     * "megabank". So we can rephrase our task as:
+     *   "there are 8 megabanks, and each megabanks contains one 4-word unit, we
+     *    want to make sure different lanes in a warp does not access different
+     *    unit addresses in the same megabank"
+     * In this terminology, matrices are in the row major format, each matrix
+     * has 8 rows, and each row has exactly one unit.
+     */
+
+    constexpr int items_per_unit = ldmatrix_cols;
+    constexpr int bytes_per_unit =
+        items_per_unit * primDataTypeSize(DataType::Half);
+    constexpr int words_per_unit = bytes_per_unit / smem_bytes_per_word;
+    constexpr int num_megabanks = smem_banks / words_per_unit;
+
+    /* In the following example, each CTA tile contains 2 rows and 3 colums of
+     * matrices, each 8x8 size:
+     *   +----------+----------+----------+
+     *   | matrix 0 | matrix 1 | matrix 2 |
+     *   +----------+----------+----------+
+     *   | matrix 3 | matrix 4 | matrix 5 |
+     *   +----------+----------+----------+
+     * The addresses of different rows in the same matrix are offset by 3 units.
+     * In this perspective, loading a matrix is a strided memory access with the
+     * following stride (in units):
+     */
+
+    // number of units per row
+    int row_stride = tile_size_y / items_per_unit;
+
+    /* So the bank conflicting problem is now converted to the following game:
+     *   I have a clock that has one pointer and `num_megabanks` ticks. I start
+     *   my game by making my pointer pointing to somewhere, and turn forward
+     *   the pointer `ldmatrix_rows` times, each time by `row_stride` ticks.
+     * This problem can be well modeled by modular arithmetic in number theory
+     * using the concept "integers modulo n" a.k.a. "Z/nZ"[1].
+     * Take n = 6 as an example, Z/6Z only has 6 elements: 0, 1, 2, 3, 4, 5.
+     * Additions and multiplications are defined in a cyclic manner:
+     *   5 + 1 = 0, 5 + 2 = 1, 5 + 3 = 2, 5 + 4 = 3, ...
+     *   2 * 1 = 2, 2 * 2 = 4, 2 * 3 = 0, 2 * 4 = 2, ...
+     * With this definition, Z is mapped to Z/nZ naturally by i -> i % n [2]
+     *
+     * It worth mention that Z/nZ is a "commutative ring", that is, we can use
+     * addition and multiplication rules just like using normal integers:
+     *   a + b = b + a, a * (b + c) = a * b + a * c, ...
+     * In short, we can reason about Z/nZ just like we are reasoning about
+     * integers, except that every number is automatically "% n".
+     *
+     * Reference:
+     * [1] https://en.wikipedia.org/wiki/Modular_arithmetic#Integers_modulo_n
+     * [2] The % is under Euclidean definition, that is -1 % 6 is 5 instead of
+     *     -1, see [The Mathematics of Integer Arithmetic] for more detail. But
+     *     we are only interested in non-negative numbers here, so there is no
+     *     need to worry about this problem
+     */
+
+    // row_stride in Z/nZ, where n is num_megabanks:
+    // assert(row_stride >= 0);
```

Comment:
@zasdfgbnm do you plan to remove these asserts, or translate them into `TORCH_INTERNAL_ASSERT` ?